File: backend/src/main/java/com/finora/common/dto/ApiResponse.java
package com.finora.common.dto;

import java.time.LocalDateTime;
import java.util.Map;

public record ApiResponse<T>(
        boolean success,
        T data,
        ApiErrorResponse error,
        Map<String, String> validationErrors,
        LocalDateTime timestamp
) {


    public static <T> ApiResponse<T> success(T data) {

        return new ApiResponse<>(
                true,
                data,
                null,
                null,
                LocalDateTime.now()
        );
    }


    public static <T> ApiResponse<T> error(
            ApiErrorResponse error
    ) {

        return new ApiResponse<>(
                false,
                null,
                error,
                null,
                LocalDateTime.now()
        );
    }


    public static <T> ApiResponse<T> validationError(
            ApiErrorResponse error,
            Map<String,String> validationErrors
    ){

        return new ApiResponse<>(
                false,
                null,
                error,
                validationErrors,
                LocalDateTime.now()
        );
    }
}

File: android/app/src/main/java/com/finora/app/data/network/ApiResponse.kt
package com.finora.app.data.network

data class ApiResponse<T>(
    val success: Boolean,
    val data: T?,
    val error: ApiErrorResponse?,
    val validationErrors: Map<String, String>?,
    val timestamp: String?
)

data class ApiErrorResponse(
    val code: String?,
    val message: String?
)
